FormParams not serializing with WebClient when charset set

With Helidon 2.5.0 we encounter the problem that sending form data with the reactive `WebClient` works, when the media type is set without a charset, but it fails when a charset is defined.

```java
webClient.post()
                .addHeader(Http.Header.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED.toString())
```

works nicely, but 
```java
webClient
                .post()
                .addHeader(Http.Header.CONTENT_TYPE,
                           MediaType.APPLICATION_FORM_URLENCODED.withCharset(StandardCharsets.UTF_8.name())
                                   .toString())
```

fails with
```
Suppressed: java.lang.IllegalStateException: Transformation failed!
		at io.helidon.media.common@2.5.0/io.helidon.media.common.MessageBodyWriterContext.marshall(MessageBodyWriterContext.java:313)
		at io.helidon.webclient@2.5.0/io.helidon.webclient.WebClientRequestBuilderImpl.submit(WebClientRequestBuilderImpl.java:432)
		at com.oracle.cx.industries.common.security.common/com.oracle.cx.industries.common.security.common.internal.oauth2.AccessTokensImpl.performOAuth2TokenRequest(AccessTokensImpl.java:188)
		at com.oracle.cx.industries.common.security.common/com.oracle.cx.industries.common.security.common.internal.oauth2.AccessTokensImpl.lambda$requestOrIssueToken$3(AccessTokensImpl.java:160)
		at io.helidon.faulttolerance@2.5.0/io.helidon.faulttolerance.DelayedTask$2.execute(DelayedTask.java:116)
		at io.helidon.faulttolerance@2.5.0/io.helidon.faulttolerance.RetryImpl.retrySingle(RetryImpl.java:91)
		at io.helidon.faulttolerance@2.5.0/io.helidon.faulttolerance.RetryImpl.lambda$retrySingle$0(RetryImpl.java:104)
		at io.helidon.common.reactive@2.5.0/io.helidon.common.reactive.SingleOnErrorResumeWith$OnErrorResumeWithSubscriber.onError(SingleOnErrorResumeWith.java:82)
		... 86 more
	Caused by: java.lang.IllegalStateException: No writer found for type: class io.helidon.common.http.FormParamsImpl. This usually occurs when the appropriate MediaSupport has not been added.
		at io.helidon.media.common@2.5.0/io.helidon.media.common.MessageBodyWriterContext.marshall(MessageBodyWriterContext.java:308)
```

--- a/common/http/src/main/java/io/helidon/common/http/MediaType.java
+++ b/common/http/src/main/java/io/helidon/common/http/MediaType.java
@@ -476,6 +476,25 @@ public final class MediaType implements AcceptPredicate<MediaType> {
     }
 
     /**
+     * Create a new {@code MediaType} instance with the same type and subtype.
+     * All parameters are trimmed. If current {@code MediaType} instance does not contain
+     * parameters, the same instance is returned.
+     *
+     * @return {@code MediaType} instance without parameters.
+     * @deprecated will be removed in 4.x version.
+     */
+    @Deprecated(since = "2.5.5", forRemoval = true)
+    public MediaType withoutParameters() {
+        if (this.parameters.isEmpty()) {
+            return this;
+        }
+        return MediaType.builder()
+                .type(this.type)
+                .subtype(this.subtype)
+                .build();
+    }
+
+    /**
      * Gets {@link Optional} value of charset parameter.
      *
      * @return Charset parameter.

--- a/media/common/src/main/java/io/helidon/media/common/FormParamsBodyReader.java
+++ b/media/common/src/main/java/io/helidon/media/common/FormParamsBodyReader.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -55,8 +55,8 @@ class FormParamsBodyReader implements MessageBodyReader<FormParams> {
     @Override
     public PredicateResult accept(GenericType<?> type, MessageBodyReaderContext context) {
         return context.contentType()
-                .filter(mediaType -> mediaType == MediaType.APPLICATION_FORM_URLENCODED
-                        || mediaType == MediaType.TEXT_PLAIN)
+                .filter(mediaType -> MediaType.APPLICATION_FORM_URLENCODED.test(mediaType)
+                        || MediaType.TEXT_PLAIN.test(mediaType))
                 .map(it -> PredicateResult.supports(FormParams.class, type))
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }
@@ -76,7 +76,7 @@ class FormParamsBodyReader implements MessageBodyReader<FormParams> {
 
     private FormParams create(String paramAssignments, MediaType mediaType, Function<String, String> decoder) {
         FormParams.Builder builder = FormParams.builder();
-        Matcher m = PATTERNS.get(mediaType).matcher(paramAssignments);
+        Matcher m = PATTERNS.get(mediaType.withoutParameters()).matcher(paramAssignments);
         while (m.find()) {
             final String key = m.group(1);
             final String value = m.group(2);

--- a/media/common/src/main/java/io/helidon/media/common/FormParamsBodyWriter.java
+++ b/media/common/src/main/java/io/helidon/media/common/FormParamsBodyWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -51,8 +51,8 @@ class FormParamsBodyWriter implements MessageBodyWriter<FormParams> {
         //User didn't have to set explicit content type. In that case set default and class filters out unsupported types.
         return context.contentType()
                 .or(() -> Optional.of(DEFAULT_FORM_MEDIA_TYPE))
-                .filter(mediaType -> mediaType == MediaType.APPLICATION_FORM_URLENCODED
-                        || mediaType == MediaType.TEXT_PLAIN)
+                .filter(mediaType -> MediaType.APPLICATION_FORM_URLENCODED.test(mediaType)
+                        || MediaType.TEXT_PLAIN.test(mediaType))
                 .map(it -> PredicateResult.supports(FormParams.class, type))
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

--- a/media/jackson/src/main/java/io/helidon/media/jackson/JacksonEsBodyStreamWriter.java
+++ b/media/jackson/src/main/java/io/helidon/media/jackson/JacksonEsBodyStreamWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -57,7 +57,6 @@ class JacksonEsBodyStreamWriter implements MessageBodyStreamWriter<Object> {
         }
         return context.contentType()
                 .or(() -> findMediaType(context))
-                .filter(mediaType -> mediaType.equals(TEXT_EVENT_STREAM_JSON) || mediaType.equals(MediaType.TEXT_EVENT_STREAM))
                 .map(it -> PredicateResult.COMPATIBLE)
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

--- a/media/jackson/src/main/java/io/helidon/media/jackson/JacksonNdBodyStreamWriter.java
+++ b/media/jackson/src/main/java/io/helidon/media/jackson/JacksonNdBodyStreamWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -56,7 +56,7 @@ class JacksonNdBodyStreamWriter implements MessageBodyStreamWriter<Object> {
         }
         return context.contentType()
                 .or(() -> findMediaType(context))
-                .filter(mediaType -> mediaType.equals(MediaType.APPLICATION_X_NDJSON))
+                .filter(MediaType.APPLICATION_X_NDJSON::test)
                 .map(it -> PredicateResult.COMPATIBLE)
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

--- a/media/jsonb/src/main/java/io/helidon/media/jsonb/JsonbEsBodyStreamWriter.java
+++ b/media/jsonb/src/main/java/io/helidon/media/jsonb/JsonbEsBodyStreamWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -57,7 +57,6 @@ class JsonbEsBodyStreamWriter implements MessageBodyStreamWriter<Object> {
         }
         return context.contentType()
                 .or(() -> findMediaType(context))
-                .filter(mediaType -> mediaType.equals(TEXT_EVENT_STREAM_JSON) || mediaType.equals(MediaType.TEXT_EVENT_STREAM))
                 .map(it -> PredicateResult.COMPATIBLE)
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

--- a/media/jsonb/src/main/java/io/helidon/media/jsonb/JsonbNdBodyStreamWriter.java
+++ b/media/jsonb/src/main/java/io/helidon/media/jsonb/JsonbNdBodyStreamWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -56,7 +56,7 @@ class JsonbNdBodyStreamWriter implements MessageBodyStreamWriter<Object> {
         }
         return context.contentType()
                 .or(() -> findMediaType(context))
-                .filter(mediaType -> mediaType.equals(MediaType.APPLICATION_X_NDJSON))
+                .filter(MediaType.APPLICATION_X_NDJSON::test)
                 .map(it -> PredicateResult.COMPATIBLE)
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

--- a/media/jsonp/src/main/java/io/helidon/media/jsonp/JsonpEsBodyStreamWriter.java
+++ b/media/jsonp/src/main/java/io/helidon/media/jsonp/JsonpEsBodyStreamWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -55,7 +55,6 @@ class JsonpEsBodyStreamWriter implements MessageBodyStreamWriter<JsonStructure> 
         }
         return context.contentType()
                 .or(() -> findMediaType(context))
-                .filter(mediaType -> mediaType.equals(TEXT_EVENT_STREAM_JSON) || mediaType.equals(MediaType.TEXT_EVENT_STREAM))
                 .map(it -> PredicateResult.COMPATIBLE)
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

--- a/media/jsonp/src/main/java/io/helidon/media/jsonp/JsonpNdBodyStreamWriter.java
+++ b/media/jsonp/src/main/java/io/helidon/media/jsonp/JsonpNdBodyStreamWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -54,7 +54,7 @@ class JsonpNdBodyStreamWriter implements MessageBodyStreamWriter<JsonStructure> 
         }
         return context.contentType()
                 .or(() -> findMediaType(context))
-                .filter(mediaType -> mediaType.equals(MediaType.APPLICATION_X_NDJSON))
+                .filter(MediaType.APPLICATION_X_NDJSON::test)
                 .map(it -> PredicateResult.COMPATIBLE)
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

--- a/media/multipart/src/main/java/io/helidon/media/multipart/MultiPartBodyWriter.java
+++ b/media/multipart/src/main/java/io/helidon/media/multipart/MultiPartBodyWriter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -48,7 +48,7 @@ public final class MultiPartBodyWriter implements MessageBodyWriter<WriteableMul
     public PredicateResult accept(GenericType<?> type, MessageBodyWriterContext context) {
         return context.contentType()
                 .or(() -> Optional.of(MediaType.MULTIPART_FORM_DATA))
-                .filter(mediaType -> mediaType == MediaType.MULTIPART_FORM_DATA)
+                .filter(MediaType.MULTIPART_FORM_DATA::test)
                 .map(it -> PredicateResult.supports(WriteableMultiPart.class, type))
                 .orElse(PredicateResult.NOT_SUPPORTED);
     }

